### PR TITLE
Initial attempt to refactor part of Optimzation.h into a distinct interface

### PR DESCRIPTION
**Pull request by sammy-tri**
No functional changes, reorg only.

Split constraints into a separate header file.

Split MathematicalProgram into its own .cpp file.

Optimization.h is still longer than I'd like, but not quite as massive as it once was.  Also MathematicalProgram has stopped reaching into private data members of a mutable object, so that's nice.

### CODE REVIEW

File: drake/core/Path.h.in [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 #include <string>
 
-namespace Drake {
+namespace {
```
Comment (RussTedrake) [inline drake/core/Path.h.in]: Why this change?

Comment (RussTedrake) [inline drake/core/Path.h.in]: I'm not particularly against it.  Just curious.

**Comment by RussTedrake** [comment]
Otherwise everything looks great to my eyes

File: drake/core/Path.h.in (hunk 1 above)
Comment (jwnimmer-tri) [inline drake/core/Path.h.in]: Google style says "Do not use unnamed namespaces in .h files." (it can break the linker).  This seems like a header file, and thus this is a defect?  Also, the change to line 5 turns the closing-brace comment (argh!) in line 10 to be lies.

Comment (sherm1) [inline drake/core/Path.h.in]: If this is just to keep an implementation detail out of the Drake namespace, a common approach is:
```c++
namespace Drake {
    namespace detail {
    }
}
```
Would that work here?

Comment (RussTedrake) [inline drake/core/Path.h.in]: I think the motivation was the opposite.. perhaps it felt redundant to reference Drake::getDrakePath().  Thank you for the style policing.  It's going to take me time to learn the new style guide.

Comment (sammy-tri) [inline drake/core/Path.h.in]: Actually, it was to get the program to link at all without duplicated symbols in every object file which had an include for Path.h.   I'll admit, it was the lazy way, I probably should have changed the code generator to emit Path.cpp instead of Path.h and produce an object file containing getDrakePath.  (an alternate option would be to #define something to @PROJECT_SOURCE_DIR@, but that didn't seem great either).

I'll see how people feel and send out, at a minimum, a fix for the closing comment if we're ok leaving an anon namespace in a header.

Comment (jpieper-tri) [inline drake/core/Path.h.in]: I would veto leaving an anonymous namespace in a header myself.  It is too easy for things to go amiss.

Comment (jwnimmer-tri) [inline drake/core/Path.h.in]: +1 to emitting Path.cpp.

Comment (sammy-tri) [inline drake/core/Path.h.in]: Ok, will go down that route.

Comment (sherm1) [inline drake/core/Path.h.in]: Emitting Path.cpp seems best to me too. There is one other option: declare the function `inline`. The language guarantees that there will only be one of those so you won't get duplicate symbol problems.

**Comment by david-german-tri** [comment]
Friendly nitpick: the file names in this pull request did not comply with the [style guide rules on naming](https://google.github.io/styleguide/cppguide.html#File_Names).  I haven't looked at any other aspect of the PR, just happened to see the files go by during `git pull`.
